### PR TITLE
Alternative implementation: Push replica labels at the end after Recv

### PR DESCRIPTION
**Pull request by fpetkovski**
This commit removes the case for a global sort on all series in the querier, and redistributes in places where series are received from stores.

This is an alternative implementation to https://github.com/thanos-io/thanos/pull/5692. 

I ran a query with this implementation against 100K series and the execution time dropped from 26s to 19s compared to main. With the new engine, the query took 9s.

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/store/proxy_heap.go [hunk 1]
```diff
@@ -256,12 +256,92 @@ func (h *ProxyResponseHeap) At() *storepb.SeriesResponse {
...
+func newUnsignedResponse(series *storepb.SeriesResponse) *dedupReadyResponse {
+	return &dedupReadyResponse{
+		signature:         0,
+		hasPushdownMarker: false,
+		SeriesResponse:    series,
+	}
+}
+
+func (m *dedupReadyResponse) LessWithoutReplicaLabels(other *dedupReadyResponse) bool {
+	if m.GetSeries() != nil && other.GetSeries() != nil {
```
Comment (GiedriusS) [inline pkg/store/proxy_heap.go:331]: It crashes for me here. Generated code looks like this:

```go
	if x, ok := m.GetResult().(*SeriesResponse_Series); ok {
		return x.Series
	}
```

It doesn't check whether `x` is not `nil` :'/ Maybe we can add a small helper:

```go

func (m *SeriesResponse) GetSeriesNotNil() *Series {
	if x, ok := m.GetResult().(*SeriesResponse_Series); ok {
		if x == nil {
			return nil
		}
		return x.Series
	}
	return nil
}
```

The problem might be somewhere else, though, that we are comparing `nil` with not `nil`. Maybe related to https://github.com/thanos-io/thanos/pull/5717?

See my comment there.

Comment (fpetkovski) [reply pkg/store/proxy_heap.go:331]: Thanks for testing this. I assume you are using the lazy strategy when it crashes?

File: pkg/store/proxy_heap.go [hunk 2]
```diff
@@ -96,11 +99,16 @@ func (d *dedupResponseHeap) Next() bool {
...
 
 		lbls := resp.GetSeries().Labels
-		lastLbls := d.responses[len(d.responses)-1].GetSeries().Labels
+		lastResponse := d.responses[len(d.responses)-1]
+		lastLbls := lastResponse.GetSeries().Labels
 
-		if labels.Compare(labelpb.ZLabelsToPromLabels(lbls), labelpb.ZLabelsToPromLabels(lastLbls)) == 0 {
+		// Unless a response comes from query pushdown, it is safe to merge
+		// chunks from replica series together into one slice.
+		labelsEqual := labels.Compare(labelpb.ZLabelsToPromLabels(lbls), labelpb.ZLabelsToPromLabels(lastLbls)) == 0
```
Comment (fpetkovski) [inline pkg/store/proxy_heap.go:107]: I would love to use `labelsEqual = resp.signature == lastResponse.signature` here, but unfortunately it breaks tests with overlapping and interleaved chunks from prometheus replicas. This was introduced and solved in:
* https://github.com/thanos-io/thanos/issues/2401
* https://github.com/thanos-io/thanos/pull/2522

Comment (fpetkovski) [reply pkg/store/proxy_heap.go:107]: I tested that change in a canary environment, and together with the new engine saw a decrease in a query from 75s to 8s, which is just insane. We have a layered querier setup, so deduplicating as soon as possible makes a massive difference.

File: pkg/store/proxy.go [hunk 3]
```diff
@@ -342,6 +351,12 @@ func (s *ProxyStore) Series(originalRequest *storepb.SeriesRequest, srv storepb.
 		}
 	}
 
+	if respHeap.UnsortedSetDetected() {
```
Comment (GiedriusS) [inline pkg/store/proxy.go:354]: This feels a bit like implementing InfoServer through the Series service. What do you think about checking `		st.SendsSortedSeries()` instead and if all are true then we could also advertise that `SendsSortedSeries() == true` in Thanos Query. And if there is at least one where `SendsSortedSeries() == false` then let's also advertise `false` in Thanos Query. 

I have been playing around with your PR a little bit and :point_up: seems like a much cleaner solution. I can try opening up a PR to it tomorrow or on Wednesday.

Comment (fpetkovski) [reply pkg/store/proxy.go:354]: Now that I read that I again, I am not sure if it is possible with the current implementation of stores to make a detection statically without inspecting the data. A user can specify any label as the replica label in a Query component. So stores need to resort data arbitrarily in the Series call if we want to have checks through info calls.

**Comment by bwplotka** [comment]
Panics on my side: 
image

**Comment by bwplotka** [comment]
let's see tomorrow (:

**Comment by GiedriusS** [comment]
> Panics on my side: image

You need to merge the latest `main` into this branch to fix this :smile:

**Comment by fpetkovski** [comment]
Might also be fixed by https://github.com/thanos-community/promql-engine/pull/55

**Comment by GiedriusS** [comment]
I think we can close this since an alternative implementation was merged for solving this problem? :smile:

**Pull request closed without merging**